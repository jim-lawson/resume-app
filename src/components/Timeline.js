import React, { useLayoutEffect, useState, useRef } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const Colors = [
  'rgb(190, 31, 14)',
  'rgb(130, 59, 134)',
  'rgb(2, 130, 174)',
  'rgb(188, 114, 27)'
]

const OneYear = 1000 * 60 * 60 * 24 * 365

const Container = styled.div`
  width: 100%;
`

const Periods = styled.div`
  margin-bottom: -10px;
  display: flex;
`

const Period = styled.div`
  border-left: 1px dashed #555;
  box-sizing: border-box;
  padding-bottom: 10px;
`

const PeriodText = styled.div`
  padding: 0 5px 5px 5px;
  font-size: 0.5em;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
`

const PeriodBar = styled.div`
  border-radius: 10px;
  height: 10px;
`

// TODO - magic number for left margin
const YearsContainer = styled.div`
  margin-left: -29px;
`

const YearsSvg = styled.svg`
  circle:nth-child(4n + 4) {
    fill: white;
  }
`

const Years = styled.ul`
  padding-left: 6px;
  display: flex;
`

const Year = styled.li`
  list-style-type: none;
  font-size: 0.5em;
`

const Timeline = props => {
  const { data, containerWidth = 0 } = props

  const containerRef = useRef()

  const [timeLineWidth, setTimeLineWidth] = useState(0)

  useLayoutEffect(() => {
    if (containerRef.current) {
      setTimeLineWidth(containerWidth || containerRef.current.offsetWidth)
    }
  }, [containerWidth])

  if (!data) {
    return null
  }

  if (!data.length) {
    console.warn('Timeline component data is empty')
    return null
  }

  const firstStartDate = data[0].startDate

  if (!firstStartDate || !(firstStartDate instanceof Date)) {
    console.warn('Timeline component has invalid start date for first event')
    return null
  }

  const lastEndDate = data[data.length - 1].endDate

  if (!lastEndDate || !(lastEndDate instanceof Date)) {
    console.warn('Timeline component has invalid end date for last event')
    return null
  }

  let ovelappingEvents = false
  data.forEach(event => {
    const otherEvents = data.filter(e => e !== event)
    otherEvents.forEach(e => {
      if (
        (e.startDate > event.startDate && e.endDate < event.endDate) ||
        (e.startDate < event.startDate && e.endDate > event.startDate)
      ) {
        ovelappingEvents = true
      }
    })
  })

  if (ovelappingEvents) {
    console.warn('Overlapping events found in timeline. This is not supported.')
    return null
  }

  const firstYear = firstStartDate.getFullYear()
  const lastYear = lastEndDate.getFullYear()
  const yearElements = []
  const yearMarkers = []

  const numYears = lastYear - firstYear + 1
  const yearWidth = parseInt(timeLineWidth / numYears)
  const circleRadius = 3
  const circleStrokeWidth = 1
  const lineStrokeWidth = 1

  for (var y = firstYear, index = 0; y < lastYear + 1; y++, index++) {
    yearElements.push(
      <Year
        key={y}
        style={{
          width: `${yearWidth}px`
        }}
      >
        <div>{y % 2 === 0 ? y : ''}</div>
      </Year>
    )

    yearMarkers.push(
      index % 2 === 0 ? (
        <circle
          key={index}
          cx={circleRadius + circleStrokeWidth + index * yearWidth}
          cy={circleRadius + circleStrokeWidth}
          r={circleRadius}
          strokeWidth={circleStrokeWidth}
          stroke="black"
        />
      ) : (
        <line
          key={index}
          x1={circleRadius + circleStrokeWidth + index * yearWidth}
          y1={0}
          x2={circleRadius + circleStrokeWidth + index * yearWidth}
          y2={circleRadius * 2 + circleStrokeWidth * 2}
          stroke="black"
        />
      )
    )
  }

  return (
    <Container ref={containerRef}>
      <Periods>
        {data.map((entry, index) => {
          let timeGap = 0
          if (index === 0) {
            timeGap = (entry.startDate.getMonth() * yearWidth) / 12
          } else {
            timeGap =
              ((entry.startDate.getTime() - data[index - 1].endDate.getTime()) /
                OneYear) *
              yearWidth
          }
          return (
            <Period
              key={index}
              style={{
                marginLeft: `${timeGap}px`,
                width: `${yearWidth *
                  ((entry.endDate.getTime() - entry.startDate.getTime()) /
                    OneYear)}px`
              }}
            >
              <PeriodText>{entry.label}</PeriodText>
              <PeriodBar
                style={{
                  background: Colors[(index + Colors.length) % Colors.length]
                }}
              />
            </Period>
          )
        })}
      </Periods>
      <YearsContainer>
        <YearsSvg
          style={{
            marginLeft: '25px',
            width: 'calc(100% - 25px)',
            height: `${circleRadius * 2 + circleStrokeWidth * 2}px`
          }}
        >
          <line
            x1={circleRadius}
            y1={circleRadius + circleStrokeWidth / 2}
            x2={timeLineWidth}
            y2={circleRadius + circleStrokeWidth / 2}
            stroke="black"
            strokeWidth={lineStrokeWidth}
          />
          {yearMarkers}
        </YearsSvg>

        {/* TODO: magic number for left margin */}
        <Years style={{ marginLeft: '14px' }}>{yearElements}</Years>
      </YearsContainer>
    </Container>
  )
}

export default Timeline

Timeline.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      startDate: PropTypes.instanceOf(Date).isRequired,
      endDate: PropTypes.instanceOf(Date).isRequired
    })
  ).isRequired,
  containerWidth: PropTypes.number
}
