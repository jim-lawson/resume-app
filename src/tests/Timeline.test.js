import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import Timeline from '../components/Timeline'

console.warn = jest.fn()
console.error = jest.fn()

afterEach(() => {
  cleanup()
  console.warn.mockClear()
  console.error.mockClear()
})

test('<Timeline /> should throw an error if no props are provided', () => {
  render(<Timeline />)
  expect(console.error).toHaveBeenCalled()
})

test('<Timeline /> should throw an error if any label is not provided', () => {
  render(<Timeline data={[{ startDate: new Date(), endDate: new Date() }]} />)
  expect(console.error).toHaveBeenCalled()
})

test('<Timeline /> should throw an error if any startDate is not provided', () => {
  render(<Timeline data={[{ label: 'foo', endDate: new Date() }]} />)
  expect(console.error).toHaveBeenCalled()
})

test('<Timeline /> should throw an error if any endDate is not provided', () => {
  render(<Timeline data={[{ label: 'foo', startDate: new Date() }]} />)
  expect(console.error).toHaveBeenCalled()
})

test('<Timeline /> should warn if data is empty', () => {
  const renderedComponent = render(<Timeline data={[]} />)
  expect(console.warn).toHaveBeenCalled()
  expect(renderedComponent.container.children.length).toEqual(0)
})

test('<Timeline /> should warn if first start date is missing', () => {
  const renderedComponent = render(
    <Timeline data={[{ label: 'foo', endDate: new Date() }]} />
  )
  expect(console.warn).toHaveBeenCalled()
  expect(renderedComponent.container.children.length).toEqual(0)
})

test('<Timeline /> should warn if last end date is missing', () => {
  const renderedComponent = render(
    <Timeline data={[{ label: 'foo', startDate: new Date() }]} />
  )
  expect(console.warn).toHaveBeenCalled()
  expect(renderedComponent.container.children.length).toEqual(0)
})

test('<Timeline /> should warn if an event partially overlaps another event', () => {
  const renderedComponent = render(
    <Timeline
      data={[
        {
          label: 'Event 1',
          startDate: new Date(1990, 0, 1),
          endDate: new Date(1991, 5, 1)
        },
        {
          label: 'Event 2',
          startDate: new Date(1990, 5, 1),
          endDate: new Date(1992, 0, 1)
        }
      ]}
    />
  )

  expect(console.warn).toHaveBeenCalled()
  expect(renderedComponent.container.children.length).toEqual(0)
})

test('<Timeline /> should warn if an event completely overlaps another event', () => {
  const renderedComponent = render(
    <Timeline
      data={[
        {
          label: 'Event 1',
          startDate: new Date(1990, 0, 1),
          endDate: new Date(1995, 5, 1)
        },
        {
          label: 'Event 2',
          startDate: new Date(1990, 5, 1),
          endDate: new Date(1992, 0, 1)
        }
      ]}
    />
  )

  expect(console.warn).toHaveBeenCalled()
  expect(renderedComponent.container.children.length).toEqual(0)
})

test('<Timeline /> should render as expected with a short time span', () => {
  const renderedComponent = render(
    <Timeline
      data={[
        {
          label: 'Event 1',
          startDate: new Date(1990, 0, 1),
          endDate: new Date(1990, 5, 1)
        },
        {
          label: 'Event 2',
          startDate: new Date(1990, 5, 1),
          endDate: new Date(1992, 0, 1)
        },
        {
          label: 'Event 3',
          startDate: new Date(1992, 0, 1),
          endDate: new Date(1993, 5, 1)
        },
        {
          label: 'Event 4',
          startDate: new Date(1993, 11, 1),
          endDate: new Date(1995, 0, 1)
        },
        {
          label: 'Event 5',
          startDate: new Date(1995, 11, 1),
          endDate: new Date(1998, 0, 1)
        }
      ]}
      containerWidth={800}
    />
  )

  expect(console.error).not.toHaveBeenCalled()
  expect(console.warn).not.toHaveBeenCalled()

  expect(renderedComponent.container.firstChild).toMatchSnapshot()
})

test('<Timeline /> should render as expected with a medium time span', () => {
  const renderedComponent = render(
    <Timeline
      data={[
        {
          label: 'Event 1',
          startDate: new Date(1990, 0, 1),
          endDate: new Date(1991, 5, 1)
        },
        {
          label: 'Event 2',
          startDate: new Date(1991, 5, 1),
          endDate: new Date(1994, 0, 1)
        },
        {
          label: 'Event 3',
          startDate: new Date(1994, 0, 1),
          endDate: new Date(1997, 5, 1)
        },
        {
          label: 'Event 4',
          startDate: new Date(1997, 11, 1),
          endDate: new Date(2002, 0, 1)
        },
        {
          label: 'Event 5',
          startDate: new Date(2002, 11, 1),
          endDate: new Date(2004, 0, 1)
        }
      ]}
    />
  )

  expect(console.error).not.toHaveBeenCalled()
  expect(console.warn).not.toHaveBeenCalled()

  expect(renderedComponent.container.firstChild).toMatchSnapshot()
})

test('<Timeline /> should render as expected with a long time span', () => {
  const renderedComponent = render(
    <Timeline
      data={[
        {
          label: 'Event 1',
          startDate: new Date(1990, 0, 1),
          endDate: new Date(1995, 5, 1)
        },
        {
          label: 'Event 2',
          startDate: new Date(1995, 5, 1),
          endDate: new Date(2001, 0, 1)
        },
        {
          label: 'Event 3',
          startDate: new Date(2001, 0, 1),
          endDate: new Date(2006, 5, 1)
        },
        {
          label: 'Event 4',
          startDate: new Date(2006, 11, 1),
          endDate: new Date(2014, 0, 1)
        },
        {
          label: 'Event 5',
          startDate: new Date(2014, 11, 1),
          endDate: new Date(2021, 0, 1)
        }
      ]}
    />
  )

  expect(console.error).not.toHaveBeenCalled()
  expect(console.warn).not.toHaveBeenCalled()

  expect(renderedComponent.container.firstChild).toMatchSnapshot()
})
