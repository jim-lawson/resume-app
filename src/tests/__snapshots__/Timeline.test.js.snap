// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timeline /> should render as expected with a long time span 1`] = `
<div
  class="sc-bdVaJa UoSJW"
>
  <div
    class="sc-bwzfXH iUXZkq"
  >
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 1
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 2
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(130, 59, 134);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 3
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(2, 130, 174);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 4
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(188, 114, 27);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 5
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
  </div>
  <div
    class="sc-EHOje gmhMvO"
  >
    <svg
      class="sc-bZQynM SLpfV"
      style="margin-left: 25px; height: 8px;"
    >
      <line
        stroke="black"
        stroke-width="1"
        x1="3"
        x2="0"
        y1="3.5"
        y2="3.5"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
    </svg>
    <ul
      class="sc-gzVnrw krluXX"
      style="margin-left: 14px;"
    >
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1990
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1992
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1994
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1996
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1998
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2000
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2002
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2004
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2006
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2008
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2010
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2012
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2014
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2016
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2018
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2020
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Timeline /> should render as expected with a medium time span 1`] = `
<div
  class="sc-bdVaJa UoSJW"
>
  <div
    class="sc-bwzfXH iUXZkq"
  >
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 1
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 2
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(130, 59, 134);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 3
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(2, 130, 174);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 4
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(188, 114, 27);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 0px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 5
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
  </div>
  <div
    class="sc-EHOje gmhMvO"
  >
    <svg
      class="sc-bZQynM SLpfV"
      style="margin-left: 25px; height: 8px;"
    >
      <line
        stroke="black"
        stroke-width="1"
        x1="3"
        x2="0"
        y1="3.5"
        y2="3.5"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="4"
        x2="4"
        y1="0"
        y2="8"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
    </svg>
    <ul
      class="sc-gzVnrw krluXX"
      style="margin-left: 14px;"
    >
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1990
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1992
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1994
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1996
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          1998
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2000
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2002
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 0px;"
      >
        <div>
          2004
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Timeline /> should render as expected with a short time span 1`] = `
<div
  class="sc-bdVaJa UoSJW"
>
  <div
    class="sc-bwzfXH iUXZkq"
  >
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 36.39543378995434px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 1
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 139.60456621004565px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 2
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(130, 59, 134);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 0px; width: 124.6365296803653px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 3
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(2, 130, 174);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 44.13059360730594px; width: 95.47397260273972px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 4
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(188, 114, 27);"
      />
    </div>
    <div
      class="sc-htpNat bhgXav"
      style="margin-left: 80.52602739726028px; width: 183.71506849315068px;"
    >
      <div
        class="sc-bxivhb bXPfCa"
      >
        Event 5
      </div>
      <div
        class="sc-ifAKCX gVKVsg"
        style="background: rgb(190, 31, 14);"
      />
    </div>
  </div>
  <div
    class="sc-EHOje gmhMvO"
  >
    <svg
      class="sc-bZQynM SLpfV"
      style="margin-left: 25px; height: 8px;"
    >
      <line
        stroke="black"
        stroke-width="1"
        x1="3"
        x2="800"
        y1="3.5"
        y2="3.5"
      />
      <circle
        cx="4"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="92"
        x2="92"
        y1="0"
        y2="8"
      />
      <circle
        cx="180"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="268"
        x2="268"
        y1="0"
        y2="8"
      />
      <circle
        cx="356"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="444"
        x2="444"
        y1="0"
        y2="8"
      />
      <circle
        cx="532"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
      <line
        stroke="black"
        x1="620"
        x2="620"
        y1="0"
        y2="8"
      />
      <circle
        cx="708"
        cy="4"
        r="3"
        stroke="black"
        stroke-width="1"
      />
    </svg>
    <ul
      class="sc-gzVnrw krluXX"
      style="margin-left: 14px;"
    >
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div>
          1990
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div>
          1992
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div>
          1994
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div>
          1996
        </div>
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div />
      </li>
      <li
        class="sc-htoDjs dvpBuU"
        style="width: 88px;"
      >
        <div>
          1998
        </div>
      </li>
    </ul>
  </div>
</div>
`;
